pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
+        for name, value in bound.arguments.items():
+            setattr(self, name, value)
+        # func may now override the attributes we just set
+        func(self, *args, **kwargs)
+    return wrapper
+
+
+class ObjectLock:
+    """Control the lock state (immutability) of a ``CCLObject``."""
+    _locked: bool = False
+    _lock_id: int = None
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(locked={self.locked})"
+
+    @property
+    def locked(self):
+        """Check if the object is locked."""
+        return self._locked
+
+    @property
+    def active(self):
+        """Check if an unlocking context manager is active."""
+        return self._lock_id is not None
+
+    def lock(self):
+        """Lock the object."""
+        self._locked = True
+        self._lock_id = None
+
+    def unlock(self, manager_id=None):
+        """Unlock the object."""
+        self._locked = False
+        if manager_id is not None:
+            self._lock_id = manager_id
+
+
+class UnlockInstance:
+    """Context manager that temporarily unlocks an immutable instance
+    of ``CCLObject``.
+
+    Parameters:
+        instance (``CCLObject``):
+            Instance of ``CCLObject`` to unlock within the scope
+            of the context manager.
+        mutate (``bool``):
+            If the enclosed function mutates the object, the stored
+            representation is automatically deleted.
+    """
+
+    def __init__(self, instance, mutate=True):
+        self.instance = instance
+        self.mutate = mutate
+        # Define these attributes for easy access.
+        self.id = id(self)
+        self.thread_lock = RLock()
+        # We want to catch and exit if the instance is not a CCLObject.
+        # Hopefully this will be caught downstream.
+        self.check_instance = isinstance(instance, CCLObject)
+        if self.check_instance:
+            self.object_lock = instance._object_lock
+
+    def __enter__(self):
+        if not self.check_instance:
+            return
+
+        with self.thread_lock:
+            # Prevent simultaneous enclosing of a single instance.
+            if self.object_lock.active:
+                # Context manager already active.
+                return
+
+            # Unlock and store the fingerprint of this context manager so that
+            # only this context manager is allowed to run on the instance.
+            self.object_lock.unlock(manager_id=self.id)
+
+    def __exit__(self, type, value, traceback):
+        if not self.check_instance:
+            return
+
+        # If another context manager is running,
+        # do nothing; otherwise reset.
+        if self.id != self.object_lock._lock_id:
+            return
+
+        with self.thread_lock:
+            # Reset `repr` if the object has been mutated.
+            if self.mutate:
+                try:
+                    delattr(self.instance, "_repr")
+                    delattr(self.instance, "_hash")
```

Comment:
This deletes the `_repr` and `_hash`, but shouldn't they be recreated? When does this happen?

Reply:
This just deletes them. This is inside the context manager `UnlockInstance`, which means that the object has been mutated. When the object is mutated, we want to discard the previously saved `_repr` and `_hash`. No need to recreate them, unless the user asks to. They are only recreated ad-hoc (i.e. when `==`, `print()`, `repr()` etc. are used).

Reply:
*Once they are recreated, how are they stored?*:
This happens inside `CCLObject`. Note that `_repr` and `_hash` are cached properties, which means that they are saved when they are computed.

Reply:
The best way to understand this is that we un-cache them using the context manager. Otherwise, they'd be stored forever and wouldn't allow us to mutate objects.